packages/fxa-settings/src/lib/relier-factory.ts

```diff
@@ -0,0 +1,162 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { Context } from "./context/interfaces/context";
+import { RelierData, Relier } from "../models/reliers/rellier"
+import { BrowserRelier } from "../models/reliers/browser-relier";
+import { OAuthRelier } from "../models/reliers/oauth-relier";
+import { RelierPairingSupplicant } from "../models/reliers/pairing-supplicant-relier";
+import { PairingAuthorityRelier } from "../models/reliers/pairing-authority-relier";
+import { UrlContext } from "./context/implementations/url-context";
+import { StorageContext } from "./context/implementations/storage-context";
+
+const DEVICE_PAIRING_SUPPLICANT_PATHNAME_REGEXP = /^\/pair\/supp/;
+const DEVICE_PAIRING_AUTHORITY_REDIRECT_URI = 'urn:ietf:wg:oauth:2.0:oob:pair-auth-webchannel';
+const SYNC_SERVICE = 'sync';
+const FX_DESKTOP_V3_CONTEXT = 'fx_desktop_v3'
+
+/**
+ * A react context compliant interface
+ */
+export interface IRelierContext {
+  relier: RelierData,
+  setRelier: (relier:RelierData) => void
+};
+
+/**
+ * Produces Reliers
+ */
+export class RelierFactory
+{
+  /**
+   * Produces a relier given the current context and flags extrapolated from that context.
+   * @param context A key value data store that holds the relier state
+   * @param flags A set of flags that help determine what kind of relier to produce. These flags are also based off the context.
+   * @returns A relier implementation.
+   */
+  static getRelier(
```

Comment:
This is intended to take the logic of creating state out of the models themselves. In theory this should making testing easier, because now we've shifted the responsibility of creation to this factory, which means we can test creation in isolation. Also, when we need to use the models in our components there is no longer any assumption about how to create state in the model itself, which should making mocking easier / more direct.